### PR TITLE
review: feature: Substitute string literals

### PR DESCRIPTION
**Pull request by pvojtechovsky**
implements #1337

### CODE REVIEW

File: src/test/java/spoon/test/template/TemplateTest.java [hunk 1]
```diff
@@ -667,7 +665,12 @@ public void createTypeFromTemplate() throws Exception {
 		assertNull(genClass.getMethod("someMethod"));
 		assertTrue(generatedIfaceMethod!=generatedClassMethod);
 		assertTrue(generatedClassMethod.isOverriding(generatedIfaceMethod));
-		
+		//contract: check replace in literals
```
Comment (monperrus) [inline src/test/java/spoon/test/template/TemplateTest.java]: what is the parameter (in `parameters`) driving this replacement?

Comment (pvojtechovsky) [inline src/test/java/spoon/test/template/TemplateTest.java]: It is the parameter implicitly added by this code
```java
class Substitution {
...
	public static <T> CtType<T> createTypeFromTemplate(String qualifiedTypeName, CtType<T> templateOfType, Map<String, Object> templateParameters) {
		final Factory f = templateOfType.getFactory();
		CtTypeReference<T> typeRef = f.Type().createReference(qualifiedTypeName);
		CtPackage targetPackage = f.Package().getOrCreate(typeRef.getPackage().getSimpleName());
		final Map<String, Object> extendedParams = new HashMap<String, Object>(templateParameters);
//--------Here is the implicit parameter added----------------
		extendedParams.put(templateOfType.getSimpleName(), typeRef);
		List<CtType<T>> generated = new SubstitutionVisitor(f, extendedParams).substitute(templateOfType.clone());
```

Comment (monperrus) [inline src/test/java/spoon/test/template/TemplateTest.java]: Implicit is worse than explicit. What if we move this code in the test? ie this responsibility to the caller?

I guess that string substitution should work  outside createTypeFromTemplate, consequently, could you add a test with a template dedicated to string substitution?

thanks

**Comment by pvojtechovsky** [comment]
OK, you are right. 
I have added explicit tests of several combinations of string literal substitution and parameter value type.
There is also tested how behaves invocation replacement - it is tested in same test, to easily see the difference in behavior.
It is ready for merge. WDYT?

**Comment by monperrus** [comment]
That's good thanks.

What I miss in SubstituteLiteralTemplate is that the template parameter are  implicit, we don't explicitly define a template parameter related to String and annotated with @Parameter. What do you think?

**Comment by pvojtechovsky** [comment]
The Template parameter is declared explicitly now. OK? ;-)

**Comment by monperrus** [comment]
thanks Pavel

**Pull request merged**